### PR TITLE
#162948836 Implement JWT Authentication

### PR DESCRIPTION
**Pull request by andrewhingah**
**What does this PR do?**
Adds unit test for JWT Authentication
Implements JWT authentication token that is sent to users upon successful registration/login

**Description of tasks to be completed**
- Refactor tests to test whether a token is sent upon successful registration/login
- Ensure users get a JWT token upon successful registration/login that they can use to make calls to all endpoints

**How should this be manually tested?**
- After cloning this repo, cd into it then `git checkout ft-jwt-authentication-162948836`
- Run `python3 manage.py runserver`
- Launch Postman and test the registration and login to see if a token is returned in both responses

**What are the relevant Pivotal Tracker stories?**
[#162948836](https://www.pivotaltracker.com/story/show/162948836)

### CODE REVIEW

File: authors/settings.py [hunk 1]
```diff
@@ -142,7 +142,7 @@
     'EXCEPTION_HANDLER': 'authors.apps.core.exceptions.core_exception_handler',
     'NON_FIELD_ERRORS_KEY': 'error',
 
-    # 'DEFAULT_AUTHENTICATION_CLASSES': (
```
Comment (SelaDanti) [inline authors/settings.py]: Do not push commented code

File: authors/apps/authentication/models.py [hunk 2]
```diff
@@ -117,4 +117,17 @@ def get_short_name(self):
...
+        """
+        date = datetime.now() + timedelta(hours=24)
+
+        payload = {
+            'email': self.email,
+            'exp': int(date.strftime('%s')),
+        }
+        
+        token = jwt.encode(payload, settings.SECRET_KEY, algorithm='HS256')
+        return token.decode()
```
Comment (kwameasiago) [inline authors/apps/authentication/models.py]: Add a line at the end of the file

**Review by kwameasiago — APPROVED** [review]
Looks Good

**Comment by ignatiusukwuoma** [comment]
Why does 3 of your commits have the same commit message?

File: authors/apps/authentication/backends.py [hunk 3]
```diff
@@ -7,4 +7,55 @@
 # from .models import User
 
 """Configure JWT Here"""
+import jwt
+import json
+from django.conf import settings
+from rest_framework import status, exceptions
```
Comment (ignatiusukwuoma) [inline authors/apps/authentication/backends.py]: django's packages first, then rest-framework's?

File: authors/apps/authentication/backends.py [hunk 4]
```diff
@@ -7,4 +7,55 @@
...
+from django.http import HttpResponse
+from rest_framework.authentication import get_authorization_header, BaseAuthentication
+from .models import User
 
+
+class JWTAuthentication(BaseAuthentication):
+
```
Comment (ignatiusukwuoma) [inline authors/apps/authentication/backends.py:12]: Add some comment (with docstrings) on this class and each of the methods

**Comment by andrewhingah** [comment]
> Why does 3 of your commits have the same commit message?

Moving forward I'll ensure to use `git commit --amend --no-edit` or different commit messages

File: authors/apps/authentication/backends.py [hunk 5]
```diff
@@ -1,10 +1,52 @@
...
+        try:
+            user = User.objects.get(
+                email=email,
+                is_active=True
+            )
+            if not user:
+                msg = "User does not exist"
+                raise exceptions.AuthenticationFailed(msg)
+        except jwt.ExpiredSignature or jwt.DecodeError or jwt.InvalidTokenError:
+            return HttpResponse({'Error': "Token is invalid"}, status="403")
```
Comment (ignatiusukwuoma) [inline authors/apps/authentication/backends.py]: Can we use the status codes from rest_framework.status?

File: authors/apps/authentication/backends.py [hunk 6]
```diff
@@ -1,10 +1,52 @@
...
+
+        return (user, token)
```
Comment (ignatiusukwuoma) [inline authors/apps/authentication/backends.py]: No need for the parenthesis, also add a blank line

File: authors/apps/authentication/models.py [hunk 7]
```diff
@@ -33,21 +34,18 @@ def create_user(self, username, email, password=None):
...
-
-      user = self.create_user(username, email, password)
-      user.is_superuser = True
-      user.is_staff = True
-      user.save()
-
-      return user
+        """
+        Create and return a `User` with superuser powers.
+        Superuser powers means that this use is an admin that can do anything
```
Comment (ignatiusukwuoma) [inline authors/apps/authentication/models.py:39]: this` user` is an admin..

File: authors/apps/authentication/models.py [hunk 8]
```diff
@@ -33,21 +34,18 @@ def create_user(self, username, email, password=None):
...
+        they want.
+        """
+        if password is None:
+            raise TypeError('Superusers must have a password.')
+            user = self.create_user(username, email, password)
```
Comment (ignatiusukwuoma) [inline authors/apps/authentication/models.py]: Why this change? If the user doesn't include a password, do we still want to create the user?

File: authors/apps/authentication/backends.py (hunk 5 above)
Comment (andrewhingah) [reply authors/apps/authentication/backends.py]: Yes. Let me change that.

File: authors/apps/authentication/models.py [hunk 9]
```diff
@@ -117,4 +115,16 @@ def get_short_name(self):
...
+        This method generates and returns a string of the token generated.
+        """
+        date = datetime.now() + timedelta(hours=24)
+
+        payload = {
+            'email': self.email,
+            'exp': int(date.strftime('%s')),
+        }
+        token = jwt.encode(payload, settings.SECRET_KEY, algorithm='HS256')
+        return token.decode()
```
Comment (ignatiusukwuoma) [inline authors/apps/authentication/models.py]: This feels like we are encoding the token... and then decoding it again. What's happening here?

Comment (andrewhingah) [reply authors/apps/authentication/models.py]: This is where we generate the token.
Supposed to be `return token`. Thanks for noting that.

File: authors/apps/authentication/models.py (hunk 7 above)
Comment (andrewhingah) [reply authors/apps/authentication/models.py:39]: Please clarify what need to be worked on here.

File: authors/apps/authentication/models.py (hunk 8 above)
Comment (andrewhingah) [reply authors/apps/authentication/models.py]: Resolved this

**Review by KMaina — APPROVED** [review]

**Review by MaggieChege — APPROVED** [review]
Looks good to me

File: authors/apps/authentication/models.py (hunk 7 above)
Comment (ignatiusukwuoma) [reply authors/apps/authentication/models.py:39]: Change use to user

File: authors/apps/authentication/views.py [hunk 10]
```diff
@@ -1,9 +1,10 @@
-from rest_framework import status
+from rest_framework import status, generics
 from rest_framework.generics import RetrieveUpdateAPIView
 from rest_framework.permissions import AllowAny, IsAuthenticated
 from rest_framework.response import Response
 from rest_framework.views import APIView
-
+from . import serializers
+from rest_framework.renderers import BrowsableAPIRenderer
```
Comment (ignatiusukwuoma) [inline authors/apps/authentication/views.py:6]: import all third-party modules/packages before yours

File: authors/apps/authentication/backends.py [hunk 11]
```diff
@@ -1,10 +1,54 @@
...
+
+        return self.authenticate_credentials(token)
+
+    def authenticate_credentials(self, token):
+        """Check that the user associated with the token
+        exists in our database"""
+        payload = jwt.decode(token, settings.SECRET_KEY)
+        email = payload['email']
+        userid = payload['id']
+        msg = {'Error': "Token mismatch", 'status': status.HTTP_401_UNAUTHORIZED}
```
Comment (ignatiusukwuoma) [inline authors/apps/authentication/backends.py]: Where are you using this msg?

Comment (andrewhingah) [reply authors/apps/authentication/backends.py]: We replaced that msg with "user does not exist", and forgot to get rid of it. Let me fix that.
